internal/controller/providerconfig/healthcheck_controller.go

```diff
@@ -104,46 +103,52 @@ func (r *HealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+	// Store the health status before the check so that we can compare
+	// with the health status after the check.
+	healthBeforeCheck := providerConfig.Status.Health
 
 	// Assume the status is Unhealthy until we can verify otherwise.
 	providerConfig.Status.Health = apisv1alpha1.HealthStatusUnhealthy
 	providerConfig.Status.Reason = ""
 	defer func() {
 		r.backendStore.SetBackendHealthStatus(req.Name, providerConfig.Status.Health)
 
-		if updateErr := r.updateConfigStatus(ctx, providerConfig, func(orig, pc *apisv1alpha1.ProviderConfig) {
-			pc.Status.Health = orig.Status.Health
-			pc.Status.Reason = orig.Status.Reason
+		if updateErr := UpdateProviderConfigStatus(ctx, r.kubeClient, providerConfig, func(pcDeepCopy, pcLatest *apisv1alpha1.ProviderConfig) {
+			pcLatest.Status.Health = pcDeepCopy.Status.Health
+			pcLatest.Status.Reason = pcDeepCopy.Status.Reason
 		}); updateErr != nil {
 			err = errors.Wrap(updateErr, err.Error())
 		}
 	}()
 
+	// Create a health check bucket on the backend if one does not already exist.
 	if err = r.bucketExists(ctx, req.Name, bucketName); err != nil {
 		if err = r.createBucket(ctx, req.Name, bucketName); err != nil {
-			r.log.Info("Failed to create bucket for health check on s3 backend", "name", providerConfig.Name, "backend", req.Name)
+			r.log.Info("Failed to create bucket for health check on s3 backend", "bucket_name", bucketName, "backend_name", providerConfig.Name)
 
-			providerConfig.Status.Reason = fmt.Sprintf("failed to create bucket: %v", err.Error())
+			providerConfig.Status.Reason = fmt.Sprintf("failed to create health check bucket: %v", err.Error())
 
 			return
 		}
 	}
 
+	// Perform the health check. By calling this function, we are implicitly updating
+	// the health status of the ProviderConfig with whatever the health check reports.
 	if err = r.doHealthCheck(ctx, providerConfig, bucketName); err != nil {
 		traces.SetAndRecordError(span, err)
-		r.log.Info("Failed to do health check on s3 backend", "name", providerConfig.Name, "backend", req.Name)
+		r.log.Info("Failed to do health check on s3 backend", "bucket_name", bucketName, "backend_name", providerConfig.Name)
 
 		providerConfig.Status.Reason = fmt.Sprintf("failed to do health check: %v", err.Error())
 
 		return
 	}
 
-	currentHealth := providerConfig.Status.Health
-
-	if currentHealth == apisv1alpha1.HealthStatusHealthy && previousHealth != currentHealth {
-		r.log.Info("Backend becomes online again", "provider", providerConfig.Name)
+	// Check if the backend is no longer healthy. In which case, we need to unpause all
+	// Bucket CRs that have buckets stored on this backend. We do this to allow these
+	// Bucket CRs be reconciled again.
+	healthAfterCheck := providerConfig.Status.Health
+	if healthBeforeCheck == apisv1alpha1.HealthStatusHealthy && healthBeforeCheck != healthAfterCheck {
+		r.log.Info("Backend is no longer healthy - unpausing all Buckets on backend to allow Observation", "backend_name", providerConfig.Name)
```

Comment:
I think this change do the opposite thing compared to the current code.
I introduced unpausing in https://github.com/linode/provider-ceph/pull/65, it aimed that paused buckets should be observed when their backends become online again.

Reply:
Good catch. Fixed